### ISSUE TITLE
panic: inconsistent list element types with module outputs

### ISSUE THREAD

**Issue by garbetjie** [hashicorp/terraform#21588]
### Terraform Version

```
Terraform v0.12.0
+ provider.google v2.7.0
+ provider.google-beta v2.7.0
+ provider.local v1.2.2
+ provider.null v2.1.2
+ provider.template v2.1.2
```

When reproducing with a stripped down example, the following versions were used:

```
Terraform v0.12.1-dev
+ provider.null v2.1.2
```

### Terraform Configuration Files

*./module1/main.tf*
```hcl
resource null_resource testing {
	count = 5
	triggers = {
		foo = "bar"
		bar = "baz"
	}
}

output triggers {
	value = null_resource.testing.*.triggers.foo
}
```

*./module2/main.tf*
```hcl
variable inputs {
  type = list
}

resource null_resource main {
	triggers = {
		foo = jsonencode(var.inputs)
	}
}
```

*./main.tf*
```hcl
module one {
  source = "./module1"
}

module two {
  source = "./module2"
  inputs = [
    { key = (["value", "value", "value", "value", "value"]) },
    { key = (module.one.triggers) },
  ]
}
```

### Debug Output
https://gist.github.com/garbetjie/12a9bb55c077529d7c943fdb889e3f0e#file-debug-log

### Crash Output
https://gist.github.com/garbetjie/12a9bb55c077529d7c943fdb889e3f0e#file-crash-log

### Expected Behavior
Output properties to be accessible and treated as an array of strings.

### Actual Behavior
Terraform crashed.

### Steps to Reproduce
1. Create module files as outlined in this bug report.
2. `terraform init`
3. `terraform plan`

### Additional Context
Initially encountered in terraform 0.12.0 installed via Homebrew. Then tested the dev version in a Docker container built off the master branch.

For what it's worth, I'm able to coerce the output to a list of strings (and thereby cause Terraform to function as expected) by joining and splitting the module's output.

Example:
```hcl
output triggers {
	value = split(",", join(",", null_resource.testing.*.triggers.foo))
}
```

**Comment by pieterbreed** [comment]
I have the same thing happening, or at least, the same error/crash behaviour. I just went through an upgrade for 0.12, which after living under a rock, I didn't expect after an innocuous `brew upgrade`...

```
$ terraform version
Terraform v0.12.3
+ provider.aws v2.17.0
+ provider.null v2.1.2
+ provider.template v2.1.2
```

On MacOS.

```
$
[crash.log](https://github.com/hashicorp/terraform/files/3339072/crash.log)
 terraform plan
panic: inconsistent list element types (cty.List(cty.DynamicPseudoType) then cty.String)

goroutine 1316 [running]:
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty.ListVal(0xc000d23a00, 0x4, 0x4, 0xc0001dd1a0, 0xc00266be00, 0x2e07480, 0x404f8f8)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/value_init.go:166 +0x4bf
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x2e075c0, 0xc0001dcf20, 0x25b4e60, 0xc0001dcf40, 0xc001956810, 0x1, 0x1, 0x4708008, 0x0, 0x100be98, ...)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert/conversion_collection.go:267 +0x54a
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2e075c0, 0xc0001dcf20, 0x25b4e60, 0xc0001dcf40, 0x0, 0x0, 0x0, 0xc0001dd120, 0xc0019567d0, 0x2e074c0, ...)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert/conversion.go:34 +0xef
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x2e075c0, 0xc0001dcf20, 0x25b4e60, 0xc0001dcf40, 0xc0019567d0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert/conversion.go:141 +0x6b
github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert.Convert(0x2e075c0, 0xc0001dcf20, 0x25b4e60, 0xc0001dcf40, 0x2e074c0, 0xc00084fc40, 0x0, 0x2e075c0, 0xc0001dcf20, 0x25b4e60, ...)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/vendor/github.com/zclconf/go-cty/cty/convert/public.go:51 +0x1af
github.com/hashicorp/terraform/terraform.(*EvalModuleCallArgument).Eval(0xc001800e00, 0x2e37d20, 0xc0018ebc70, 0x2, 0x2, 0xc001d1e000, 0x52)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval_variable.go:162 +0x160
github.com/hashicorp/terraform/terraform.EvalRaw(0x2dc35c0, 0xc001800e00, 0x2e37d20, 0xc0018ebc70, 0x3d, 0x0, 0x0, 0x3d)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc0040af6b0, 0x2e37d20, 0xc0018ebc70, 0x2, 0x2, 0xc001b23b10, 0x103d494)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/hashicorp/terraform/terraform.EvalRaw(0x2dc3600, 0xc0040af6b0, 0x2e37d20, 0xc0018ebc70, 0x14, 0xbf3d9de60e161828, 0x2acb2b46e, 0x3d)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc0004a2720, 0x2e37d20, 0xc0018ebc70, 0x2, 0x2, 0x1acf2d5, 0x2dc3760)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x2dc3740, 0xc0004a2720, 0x2e37d20, 0xc0018ebc70, 0x266ece0, 0x3f05405, 0x25d8620, 0xc001479be0)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.Eval(0x2dc3740, 0xc0004a2720, 0x2e37d20, 0xc0018ebc70, 0xc0004a2720, 0x2dc3740, 0xc0004a2720, 0x0)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x281daa0, 0xc001156540, 0x0, 0x0, 0x0)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/terraform/graph.go:90 +0xf40
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc001076f80, 0x281daa0, 0xc001156540, 0xc0011a6f40)
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x353
created by github.com/hashicorp/terraform/dag.(*Walker).Update
        /private/tmp/terraform-20190625-7012-173qxtw/terraform-0.12.3/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xa9b

```

**Comment by n3ph** [comment]
Same here in a extremely more complex code base... -> https://github.com/Reservix/terraform-module-alb

```
$ terraform version
Terraform v0.12.4
+ provider.aws v2.18.0
```
On Debian Testing

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"config":cty.Map(cty.String), "type":cty.String}) then cty.Object(map[string]cty.Type{"config":cty.Map(cty.DynamicPseudoType), "type":cty.String}))

goroutine 201 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc00002dc40, 0x2, 0x2, 0xc000e0c960, 0xc00000d6e0, 0x2, 0x2)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/value_init.go:166 +0x4bf
github.com/zclconf/go-cty/cty/convert.conversionTupleToList.func2(0x2207100, 0xc00000d2e0, 0x19ba860, 0xc00000d300, 0xc00000d6e0, 0x2, 0x2, 0x1, 0xc000e6d380, 0x409828, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/conversion_collection.go:267 +0x54a
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2207100, 0xc00000d2e0, 0x19ba860, 0xc00000d300, 0xc00088b460, 0x1, 0x1, 0x19ba860, 0xc00000d300, 0xc00088b401, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/conversion.go:34 +0xef
github.com/zclconf/go-cty/cty/convert.conversionObjectToObject.func1(0x2207080, 0xc00000e370, 0x1a80340, 0xc000e0ce70, 0xc00088b460, 0x1, 0x1, 0x7f2c9d5906d0, 0x0, 0x40c0d8, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/conversion_object.go:65 +0x49c
github.com/zclconf/go-cty/cty/convert.getConversion.func1(0x2207080, 0xc00000e370, 0x1a80340, 0xc000e0ce70, 0x0, 0x0, 0x0, 0xc00000d6a0, 0xc00088b450, 0x2207080, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/conversion.go:34 +0xef
github.com/zclconf/go-cty/cty/convert.retConversion.func1(0x2207080, 0xc00000e370, 0x1a80340, 0xc000e0ce70, 0xc00088b450, 0x0, 0x0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/conversion.go:141 +0x6b
github.com/zclconf/go-cty/cty/convert.Convert(0x2207080, 0xc00000e370, 0x1a80340, 0xc000e0ce70, 0x2207080, 0xc00041a3a8, 0x0, 0x2207080, 0xc00000e370, 0x1a80340, ...)
	/opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.1-0.20190708163926-19588f92a98f/cty/convert/public.go:51 +0x1af
github.com/hashicorp/terraform/terraform.(*EvalModuleCallArgument).Eval(0xc0004fc1c0, 0x2238c40, 0xc0010561a0, 0x2, 0x2, 0xc000dfa000, 0x42)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval_variable.go:162 +0x160
github.com/hashicorp/terraform/terraform.EvalRaw(0x21c5920, 0xc0004fc1c0, 0x2238c40, 0xc0010561a0, 0x2d, 0x0, 0x0, 0x2d)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalOpFilter).Eval(0xc001538120, 0x2238c40, 0xc0010561a0, 0x2, 0x2, 0xc000c94b38, 0x42b7e1)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval_filter_operation.go:37 +0x4c
github.com/hashicorp/terraform/terraform.EvalRaw(0x21c5960, 0xc001538120, 0x2238c40, 0xc0010561a0, 0x14, 0xbf435ec01ccda1d2, 0xf7acbc97, 0x2d)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc000320d40, 0x2238c40, 0xc0010561a0, 0x2, 0x2, 0xed09a5, 0x21c5ac0)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:20 +0xfd
github.com/hashicorp/terraform/terraform.EvalRaw(0x21c5aa0, 0xc000320d40, 0x2238c40, 0xc0010561a0, 0x1a70da0, 0x31c9105, 0x19dde20, 0xc000d8cf50)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval.go:57 +0x131
github.com/hashicorp/terraform/terraform.Eval(0x21c5aa0, 0xc000320d40, 0x2238c40, 0xc0010561a0, 0xc000320d40, 0x21c5aa0, 0xc000320d40, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/eval.go:35 +0x4d
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x1c20800, 0xc00002cdc0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/graph.go:90 +0xf40
github.com/hashicorp/terraform/dag.(*Walker).walkVertex(0xc00065d780, 0x1c20800, 0xc00002cdc0, 0xc00002c580)
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/dag/walk.go:392 +0x353
created by github.com/hashicorp/terraform/dag.(*Walker).Update
	/opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/dag/walk.go:314 +0xa9b



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

Crash log: https://github.com/Reservix/terraform-module-alb/blob/master/test/crash.log

Removing: Without the first element in actions list [1] everything works as expected..

[1] https://github.com/Reservix/terraform-module-alb/blob/master/test/main.tf#L190-L201

**Comment by tesharp** [comment]
Have similar problem with input variables

```bash
$ terraform version
Terraform v0.12.6
```

Module input variable defined as an object with list of any:

```terraform
variable "input" {
  type = list(object({ something = list(any) }))
}
```

Trying to send empty list once and for another list item some other input.

```terraform
module "test" {
  source = "..."

  input = [
    { something = [] },
    { something = [{ test = "test" }] }
  ]
}
```

Error message:

```
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"something": cty.List(cty.DynamicPseudoType)}) then cty.Object(map[string]cty.Type{"something": cty.List(cty.Object(map[string]cty.Type{"test": cty.String}))}))
```

**Comment by thomassampson** [comment]
Also experiencing the same issue,

**Comment by sahilkhanna9** [comment]
Experiencing the same Issue with the Azure Provider. 

module->NetworkSecurityGroups/main.tf
resource "azurerm_network_security_group" "abc" {
  name                                     = var.name
  location                                 = var.location
  resource_group_name           = var.resource_group_name

  dynamic "security_rule" {
    for_each = [for s in var.custom_security_rule:{
      name                       = s.name
      priority                   = s.priority
      direction                  = s.direction
      access                     = s.access
      protocol                   = s.protocol
      source_port_range          = s.source_port_range
      destination_port_ranges     = s.destination_port_ranges
      source_address_prefix      = s.source_address_prefix
      destination_address_prefix = s.destination_address_prefix
      description                = s.description
    }]
    content {
      name                       = security_rule.value.name
      priority                   = security_rule.value.priority
      direction                  = security_rule.value.direction
      access                     = security_rule.value.access
      protocol                   = security_rule.value.source_port_range
      source_port_range          = security_rule.value.source_port_range
      destination_port_ranges     = security_rule.value.destination_port_ranges
      source_address_prefix      = security_rule.value.source_address_prefix
      destination_address_prefix = security_rule.value.destination_address_prefix
      description                = security_rule.value.description
    }
  }
}  




main.tf
module "abc_nsg" {
  source = "./modules/NetworkSecurityGroups"

  name                         = .abc
  resource_group_name   = abcrg
  location                     = location
  
  custom_security_rule         = [{
    name                       = "local_subnet"
    priority                   = 100
    direction                  = "Inbound"
    access                     = "Allow"
    protocol                   = "*"
    source_port_range          = "*"
    destination_port_ranges    = ["5985", "5986"]
    source_address_prefix      = var.subnet_address
    destination_address_prefix = var.subnet_address
    description                = ""
    }]
}

Output: 
terraform plan
panic: inconsistent map element types (cty.String then cty.Tuple([]cty.Type{cty.String, cty.String}))
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
--------------------------------------------------------------------------------------------------

If I use destination_port_range and pass a single string instead of using destination_port_ranges and a list, it just works fine. Issue is arising when passing a list.

**Comment by LittleMikeDev** [comment]
Also had this same error message when trying to build an environment from scratch. I never managed to pinpoint the configuration that caused it because when I uncommented and applied bits of configuration in small chunks, everything magically worked.

I know this doesn't help diagnose the issue, but if anyone else is getting totally blocked by this error, you might have some luck just gradually planning/applying your configuration in small chunks :-S

**Comment by jspiro** [comment]
I'm getting this too, but the difference between each list element is simply that some keys are set to null in one, and those keys are populated with a map in the other–the purpose was to be able to use for_each in a dynamic block and only create the block when it's specified.

Any workarounds?

**Comment by nekobrain** [comment]
> I'm getting this too, but the difference between each list element is simply that some keys are set to null in one, and those keys are populated with a map in the other–the purpose was to be able to use for_each in a dynamic block and only create the block when it's specified.
> 
> Any workarounds?

Uhm... have you tried not including the keys at all? I have a dynamic block which gets created only if a variable gets passed to the module, otherwise TF ignores it.

**Comment by DrMerlin** [comment]
same here- maddening bug.
We were able to work around it by defining the exact shape of the map that gets passed into the module.  I.e. number and type of all key/pairs.

**Comment by gaijin03** [comment]
I believe I'm hitting this same issue. I've narrowed it down to a simple test.

```
$ cat test.tf
variable "test" {
  type = list(object({
    name  = string,
    labels = any
  }))
}

output "test_out" {
  value = var.test
}

$ cat io.tfvars
test = [
  {
    name  = "blah1"
    labels = {}
  },
  {
    name  = "blah2"
    labels = {
      var1 = "val1",
      var2 = "var2"
    }
  }
]
```

```
$ terraform apply -var-file=io.tfvars
panic: inconsistent list element types (cty.Object(map[string]cty.Type{"labels":cty.EmptyObject, "name":cty.String}) then cty.Object(map[string]cty.Type{"labels":cty.Object(map[string]cty.Type{"var1":cty.String, "var2":cty.String}), "name":cty.String}))
...
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

It crashes when "labels" doesn't have the same keys.
e.g. both of these don't crash.
```
test = [
  {
    name  = "blah1"
    labels = {}
  },
  {
    name  = "blah2"
    labels = {}
  }
]
```

```
test = [
  {
    name  = "blah1"
    labels = {
      var1 = "val1",
      var2 = "var2"
    }
  },
  {
    name  = "blah2"
    labels = {
      var1 = "val1",
      var2 = "var2"
    }
  }
]
```

A workaround is to not define the type of the variable.

**Comment by next-jesusmanuelnavarro** [comment]
Same issue here, this time when a variable's value got updated from boolean to object:

on variables.tf:
```
variable "with_var" {
  type = any
}
```
and a vars.tfvars file.
* First terraform went ok when...
```
with_var = false
 ```
*  ...then, terraform panic'ed when `with_var` was updated from the boolen *false* value to an object:
```
  with_var = {
    some_key = "some_value"
    other_key  = "other_value"
  }
```

```
panic: inconsistent list element types (..., "with_var":cty.Bool})
```

**Comment by danieldreier** [comment]
I have reproduced this on 0.12.25 using @gaijin03's simplified test case. Thanks for writing that up!

**Comment by pankajmt** [comment]
can you please expand on "A workaround is to not define the type of the variable." @gaijin03 ?

**Comment by gaijin03** [comment]
> can you please expand on "A workaround is to not define the type of the variable." @gaijin03 ?

@pankajmt In my test.tf example, if you comment out/remove "labels = any" then it won't crash.

**Comment by pankajmt** [comment]
but then you cannot access configuration for labels, right? the workaround i have understood to access labels is to define the exact shape of labels and have it specified everywhere!

**Comment by mildwonkey** [comment]
Thanks for reporting this issue, and to everyone who helped with reproduction cases. The panic has been fixed in the upstream library and this will be included in the upcoming 0.13 release. 

Here's the output from @gaijin03 's test case using terraform built from master:

<img width="718" alt="Screen Shot 2020-05-18 at 2 14 26 PM" src="https://user-images.githubusercontent.com/6210214/82245962-e5602a80-9911-11ea-8346-bf071aea9c42.png">
